GHC 9 compatibility

When building `moat` with GHC 9, I get the following error:

```
moat                             > [3 of 6] Compiling Moat.Pretty.Kotlin
moat                             > [4 of 6] Compiling Moat.Class
moat                             > [5 of 6] Compiling Moat
moat                             >
moat                             > /tmp/stack-46a8b3125826381c/moat-0.1/src/Moat.hs:322:21: error:
moat                             >     • Expecting one more argument to ‘TyVarBndr’
moat                             >       Expected a type, but ‘TyVarBndr’ has kind ‘* -> *’
moat                             >     • In the type ‘[TyVarBndr]’
moat                             >       In the definition of data constructor ‘NewtypeInfo’
moat                             >       In the data declaration for ‘NewtypeInfo’
moat                             >     |
moat                             > 322 |     newtypeVars :: [TyVarBndr],
moat                             >     |                     ^^^^^^^^^
```

--- a/moat.cabal
+++ b/moat.cabal
@@ -1,6 +1,6 @@
 cabal-version: 1.12
 
--- This file has been generated from package.yaml by hpack version 0.34.3.
+-- This file has been generated from package.yaml by hpack version 0.34.4.
 --
 -- see: https://github.com/sol/hpack
 
@@ -34,22 +34,27 @@ library
       Paths_moat
   hs-source-dirs:
       src
-  default-extensions: LambdaCase RecordWildCards
+  default-extensions:
+      LambdaCase
+      RecordWildCards
   ghc-options: -Wall
+  build-tool-depends:
+      hspec-discover:hspec-discover
   build-depends:
-      base >=4.11 && <4.15
-    , bytestring >=0.10 && <0.11
-    , case-insensitive >=1.2 && <1.3
+      base >=4.11 && <4.16
+    , bytestring ==0.10.*
+    , case-insensitive ==1.2.*
     , containers >=0.5.9 && <0.7
-    , mtl >=2.2 && <2.3
+    , mtl ==2.2.*
     , primitive >=0.6.4 && <0.8
-    , template-haskell >=2.11 && <2.17
-    , text >=1.2 && <1.3
+    , template-haskell >=2.11 && <2.18
+    , text ==1.2.*
     , th-abstraction >=0.3 && <0.5
+    , th-compat >=0.1.0 && <0.2
     , time >=1.8 && <1.11
-    , unordered-containers >=0.2 && <0.3
-    , uuid-types >=1.0 && <1.1
-    , vector >=0.12 && <0.13
+    , unordered-containers ==0.2.*
+    , uuid-types ==1.0.*
+    , vector ==0.12.*
   default-language: Haskell2010
 
 test-suite spec
@@ -79,23 +84,33 @@ test-suite spec
   hs-source-dirs:
       test
       src
-  default-extensions: LambdaCase RecordWildCards AllowAmbiguousTypes DataKinds ScopedTypeVariables TemplateHaskell TypeApplications
+  default-extensions:
+      LambdaCase
+      RecordWildCards
+      AllowAmbiguousTypes
+      DataKinds
+      ScopedTypeVariables
+      TemplateHaskell
+      TypeApplications
   ghc-options: -Wall -Wno-unused-top-binds
+  build-tool-depends:
+      hspec-discover:hspec-discover
   build-depends:
-      base >=4.11 && <4.15
-    , bytestring >=0.10 && <0.11
-    , case-insensitive >=1.2 && <1.3
+      base >=4.11 && <4.16
+    , bytestring ==0.10.*
+    , case-insensitive ==1.2.*
     , containers >=0.5.9 && <0.7
     , hspec
     , hspec-discover
     , hspec-golden
-    , mtl >=2.2 && <2.3
+    , mtl ==2.2.*
     , primitive >=0.6.4 && <0.8
-    , template-haskell >=2.11 && <2.17
-    , text >=1.2 && <1.3
+    , template-haskell >=2.11 && <2.18
+    , text ==1.2.*
     , th-abstraction >=0.3 && <0.5
+    , th-compat >=0.1.0 && <0.2
     , time >=1.8 && <1.11
-    , unordered-containers >=0.2 && <0.3
-    , uuid-types >=1.0 && <1.1
-    , vector >=0.12 && <0.13
+    , unordered-containers ==0.2.*
+    , uuid-types ==1.0.*
+    , vector ==0.12.*
   default-language: Haskell2010

--- a/package.yaml
+++ b/package.yaml
@@ -16,19 +16,23 @@ default-extensions:
   - RecordWildCards
 
 dependencies:
-  - base >= 4.11 && < 4.15
+  - base >= 4.11 && < 4.16
   - bytestring >= 0.10 && < 0.11
   - case-insensitive >= 1.2 && < 1.3
   - containers >= 0.5.9 && < 0.7
   - mtl >= 2.2 && < 2.3
   - primitive >= 0.6.4 && < 0.8
-  - template-haskell >= 2.11 && < 2.17
+  - template-haskell >= 2.11 && < 2.18
   - text >= 1.2 && < 1.3
   - th-abstraction >= 0.3 && < 0.5
+  - th-compat >= 0.1.0 && < 0.2
   - time >= 1.8 && < 1.11
   - unordered-containers >= 0.2 && < 0.3
   - uuid-types >= 1.0 && < 1.1
   - vector >= 0.12 && < 0.13
+
+build-tools:
+  - hspec-discover
 
 library:
   source-dirs: src

--- a/src/Moat.hs
+++ b/src/Moat.hs
@@ -1,4 +1,5 @@
 {-# LANGUAGE BangPatterns #-}
+{-# language PatternSynonyms #-}
 {-# LANGUAGE CPP #-}
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE DeriveGeneric #-}
@@ -91,8 +92,9 @@ import Data.Proxy (Proxy (..))
 import qualified Data.Text as TS
 import Data.Void (Void)
 import GHC.TypeLits (KnownSymbol, Symbol, symbolVal)
-import Language.Haskell.TH hiding (stringE, tupE)
-import Language.Haskell.TH.Datatype
+import qualified Language.Haskell.TH
+import Language.Haskell.TH hiding (stringE, tupE, TyVarBndr(..))
+import Language.Haskell.TH.Datatype hiding (TyVarBndr (..))
 import qualified Language.Haskell.TH.Syntax as Syntax
 import Moat.Class
 import Moat.Pretty.Kotlin (prettyKotlinData)
@@ -100,6 +102,17 @@ import Moat.Pretty.Swift (prettySwiftData)
 import Moat.Types hiding (newtypeName)
 import qualified Moat.Types
 import Prelude hiding (Enum (..))
+import Language.Haskell.TH.Syntax.Compat
+
+#if MIN_VERSION_template_haskell(2,17,0)
+type TyVarBndr = Language.Haskell.TH.TyVarBndr ()
+
+pattern PlainTV n = Language.Haskell.TH.PlainTV n ()
+
+pattern KindedTV n k <- Language.Haskell.TH.KindedTV n _ k
+#else
+type TyVarBndr = Language.Haskell.TH.TyVarBndr
+#endif
 
 -- Used internally to reflect polymorphic type
 -- variables into TH, then reify them into 'Poly'.
@@ -222,7 +235,7 @@ getTags ::
   MoatM ([Exp], [Dec])
 getTags parentName ts = do
   let b = length ts > 1
-  disambiguate <- lift [||b||]
+  disambiguate <- lift $ examineCode [||b||]
   foldlM
     ( \(es, ds) n -> do
         NewtypeInfo {..} <- reifyNewtype n
@@ -1209,10 +1222,20 @@ canRealiseKindStar = \case
   _ -> NotKindStar
 
 -- discard the kind signature from a TyVarBndr.
-tyVarBndrNoSig :: TyVarBndr -> TyVarBndr
+tyVarBndrNoSig
+#if MIN_VERSION_template_haskell(2,17,0)
+  :: Language.Haskell.TH.TyVarBndr a
+  -> Language.Haskell.TH.TyVarBndr Specificity
+tyVarBndrNoSig = \case
+  Language.Haskell.TH.PlainTV n x -> Language.Haskell.TH.PlainTV n SpecifiedSpec
+  Language.Haskell.TH.KindedTV n _ _k -> Language.Haskell.TH.PlainTV n SpecifiedSpec
+#else
+  :: TyVarBndr
+  -> TyVarBndr
 tyVarBndrNoSig = \case
   PlainTV n -> PlainTV n
   KindedTV n _k -> PlainTV n
+#endif
 
 -- fully applies a type constructor to its
 -- type variables

